transports/jms/src/main/java/org/mule/transport/jms/JmsMessageDispatcher.java

```diff
@@ -95,186 +98,310 @@ protected boolean isDisableTemporaryDestinations()
-                    listener.release();
-                    Message result = listener.getMessage();
-                    if (result == null)
-                    {
-                        logger.debug("No message was returned via replyTo destination");
-                        return createNullMuleMessage();
-                    }
-                    else
-                    {
-                        return createMessageWithJmsMessagePayload(result);
-                    }
+                    internalNonBlockingSendAndReceive(session, producer, consumer, replyTo, jmsMessage, topic, ttl, priority, persistent, transacted, timeout, completionHandler);
+                    delayedCleanup = true;
+                    return null;
                 }
                 else
                 {
-                    if (logger.isDebugEnabled())
+                    try
                     {
-                        logger.debug("Waiting for return event for: " + timeout + " ms on " + replyTo);
-                    }
+                        if (topic)
+                        {
+                            return internalBlockingSendAndAwait(consumer, producer, replyTo, jmsMessage, topic, ttl, priority, persistent, timeout);
 
-                    Message result = consumer.receive(timeout);
-                    if (result == null)
-                    {
-                        logger.debug("No message was returned via replyTo destination " + replyTo);
-                        return createNullMuleMessage();
+                        }
+                        else
+                        {
+                            return internalBlockingSendAndReceive(producer, consumer, replyTo, jmsMessage, topic, ttl, priority, persistent, timeout);
+                        }
                     }
-                    else
+                    finally
                     {
-                        return createMessageWithJmsMessagePayload(result);
+                        closeConsumer(session, consumer, replyTo);
                     }
                 }
             }
             else
             {
-                // In this case a response was never expected so we return null and not NullPayload.
-                // This generally happens when dispatch is used for an asynchronous endpoint but can also occur when send() is used 
-                // and disableTempDestinations is set.
-                return returnOriginalMessageAsReply ? createMuleMessage(msg) : null;
+                return internalSend(producer, jmsMessage, topic, ttl, priority, persistent);
             }
         }
         finally
         {
             connector.closeQuietly(producer);
-            connector.closeQuietly(consumer);
+            if(!delayedCleanup)
+            {
+                closeSession(session, transacted);
+            }
+        }
+    }
+
+    private String getCorrelationId(MuleEvent event)
+    {
+        if (event.getFlowConstruct() != null)
```

Comment:
CorrelationId and non-blocking are unrelated. In the previous PR this method was also private.  WMQ overrides the method that creates the JMS correlationID, but does not need to override the creation of the Mule correlationID which is deliberately independent of any specific transport implementation.